udev-117/extras/volume_id/vol_id.8
.\"     Title: vol_id
.\"    Author:
.\" Generator: DocBook XSL Stylesheets v1.73.1 <http://docbook.sf.net/>
.\"      Date: March 2006
.\"    Manual: vol_id
.\"    Source: volume_id
.\"
.TH "VOL_ID" "8" "March 2006" "volume_id" "vol_id"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
vol_id - probe filesystem type and read label and uuid
.SH "SYNOPSIS"
.HP 7
\fBvol_id\fR [\fB\-\-export\fR] [\fB\-\-type\fR] [\fB\-\-label\fR] [\fB\-\-label\-raw\fR] [\fB\-\-uuid\fR] [\fB\-\-skip\-raid\fR] [\fB\-\-probe\-all\fR] [\fB\-\-help\fR] [\fIdevice\fR]
.SH "DESCRIPTION"
.PP
\fBvol_id\fR
is usually called from a udev rule, to provide udev with the filesystem type, the label and the uuid of a volume\. It supports most of the common filesystem formats and detects various raid setups to prevent the recognition of raid members as a volume with a filesystem\.
.SH "OPTIONS"
.PP
\fB\-\-export\fR
.RS 4
Print all values in key/value format to import them into the environment\.
.RE
.PP
\fB\-\-type\fR
.RS 4
Print the filesystem type\.
.RE
.PP
\fB\-\-label\fR
.RS 4
Print the safe version of volume label suitable for use as filename\.
.RE
.PP
\fB\-\-label\-raw\fR
.RS 4
Print the raw volume label\.
.RE
.PP
\fB\-\-uuid\fR
.RS 4
Print the uuid of a volume\.
.RE
.PP
\fB\-\-skip\-raid\fR
.RS 4
Skip detection of raid metadata\.
.RE
.PP
\fB\-\-probe\-all\fR
.RS 4
Probe for all types and print all matches\.
.RE
.PP
\fB\-\-help\fR
.RS 4
Print usage\.
.RE
.SH "ENVIRONMENT"
.PP
\fBUDEV_LOG\fR
.RS 4
Set the syslog priority\.
.RE
.SH "EXIT STATUS"
.PP
\fBvol_id\fR
will only return successful if the value asked for is not empty\. All trailing whitespace will be removed, spaces replaced by underscore and slashes ignored\.
.SH "AUTHOR"
.PP
Written by Kay Sievers
<kay\.sievers@vrfy\.org>
.SH "SEE ALSO"
.PP
\fBudev\fR(7)


udev-117/ChangeLog
Summary of changes from v116 to v117
============================================

Dan Nicholson (2):
      extras: ignore built and generated files
      volume_id: create relative symlink when $(libdir) = $(usrlibdir)

Kay Sievers (15):
      usb_id: fail if vendor/product can not be retrieved
      rules: SUSE update
      firmware: do not print error if logger is missing
      volume_id: vfat - allow all possible sector sizes
      volume_id: LUKS - export version
      volume_id: ntfs - rely on valid master file table
      volume_id: bump version
      udevinfo: exclude "uevent" file from --attribute-walk
      udevadm: merge all udev tools into a single binary
      udevadm: accept command as option, like --help, --version
      udevadm: add info option --device-id-of-file=<file>
      Makefile: fix bogus version number than got committed
      udevadm: also return major==0 results for --device-id-of-file
      man: udevd.8 - remove udevcontrol section
      udevadm: control - allow command to be passed as option

MUNEDA Takahiro (1):
      man: fix udevadm.8 typo

Matthias Schwarzott (2):
      firmware: remove hardcoded path to logger
      rules: Gentoo update

VMiklos (1):
      rules: Frugalware update


Summary of changes from v115 to v116
============================================

Bryan Kadzban (1):
      rules: fix typos

Harald Hoyer (3):
      check line length after comment check and whitespace strip
      only install *.rules
      remove extra space from udevinfo symlink output

Kay Sievers (29):
      rules: fix two trivial typos
      rules: random and urandom are 0666
      rules: add REMOVE_CMD rule
      track "move" events to rename database and failed files
      rules: Gentoo update
      rules: add i2o driver rule
      man: recreate man pages
      volume_id: fix linux_raid metadata version 1.0 detection
      add $name substitution
      do not delete the device node with ignore_remove, but handle the event
      print warning for invalid TEST operations
      rules: do not delete /lib/udev/devices/ nodes on "remove"
      rules: remove broken nvram group assignment without any permission
      add /dev/rtc symlink if new rtc drivers are used
      increase WAIT_FOR_SYSFS timeout to 10 seconds
      rules: put bsd nodes in /dev/bsd/ directory
      path_id: fix for stacked class devices
      ignore device node names while restoring symlinks from the stack
      use SEQNUM in /dev/.udev/queue/ instead of devpath
      rules: add memstick module loading
      udevinfo: simplify symlink printing logic
      prevent wrong symlink creation if database disagress with current rules
      fix wrong variable used in logged string
      update README
      rule_generator: move all policy from write_net_rules to the rules file
      rules: call usb_id only for SUBSYSTEMS=="usb"
      rules: split out and fix persistent tape rules
      fix debug output string
      rule_generator: always match netif type in generated rule

Matthias Schwarzott (3):
      rules: Gentoo update
      rules: Gentoo update
      rules: Gentoo update

Michael Morony (1):
      set buffer size if strlcpy/strlcat indicate truncation

maximilian attems (1):
      correct includes in udev_selinux.c


Summary of changes from v114 to v115
============================================

Harald Hoyer (1):
      rules: fix typo in 80-drivers.rules

Kay Sievers (15):
      rules: add default rules
      rules: update SUSE rules
      rules: add packages rules
      rules: add ia64 rules
      rules: move md-raid rules to packages dir
      rules: run vol_id only for partitions
      rules: update Fedora rules
      edd_id: move persistent rules to its own file
      accept relative path for TEST
      rules: add iowarrior rule
      volume_id: fix sqashfs detection
      do not ignore dynamic rule if it is the last one in the list
      rule_generator: fix wrong DRIVERS!= logic
      rules: update Fedora
      Makefile: install default rules

Marco d'Itri (3):
      rules_generator: remove policy from write_cd_rules
      rules_generator: fix write_cd_rules when similar names exist in the root directory
      rules: Debian update


Summary of changes from v113 to v114
============================================

Hannes Reinecke (3):
      collect: extra to synchronize actions across events
      add $driver subtitution
      rules_generator: add S/390 persistent network support

Kay Sievers (24):
      rules_generator: remove executable flag from include file
      always unlink temporary file before creating new one
      rules: SUSE update
      volume_id: ext4 detection
      udevtrigger: allow to specify action string
      add option to RUN key to ignore the return value of the program
      use global udev_log variable instead of parameter in run_program
      add udev_rules_run() to handle RUN list
      move udev_utils_run.c into udev_rules.c
      rules: SUSE update
      name_list: rename loop_name -> name_loop
      handle dynamic rules created in /dev/.udev/rules.d/
      allow SYMLINK== match
      libvolume_id: use /usr/$libdir in pc file
      Makefile: add --as-needed flag to ld
      restore behavior of NAME==
      rules_generator: remove "installation" function
      udevtrigger: trigger "driver" events
      rules: update SUSE
      rules: Fedora update
      rules: add "do not edit" comment
      rules: Fedora update
      rules_generator: skip random MAC addresses
      write changed network interface names to the kernel log

Matthias Schwarzott (3):
      rules: Gentoo update
      fix inotify to work not only once
      rules: Gentoo update

Richard Hughes (1):
      Makefile: add "make dist" for nightly snapshots


Summary of changes from v112 to v113
============================================

David Zeuthen (1):
      vol_id: do not fail if unable to drop privileges

Kay Sievers (12):
      add missing ChangeLog

udev-117/extras/volume_id/vol_id.c
/*
 * vol_id - read filesystem label and uuid
 *
 * Copyright (C) 2005-2006 Kay Sievers <kay.sievers@vrfy.org>
 *
 *	This program is free software; you can redistribute it and/or modify it
 *	under the terms of the GNU General Public License as published by the
 *	Free Software Foundation version 2 of the License.
 *
 *	This program is distributed in the hope that it will be useful, but
 *	WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *	General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License along
 *	with this program; if not, write to the Free Software Foundation, Inc.,
 *	675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#ifndef _GNU_SOURCE
#define _GNU_SOURCE 1
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>
#include <pwd.h>
#include <grp.h>
#include <getopt.h>
#include <fcntl.h>
#include <sys/ioctl.h>

#include "../../udev.h"
#include "lib/libvolume_id.h"

#define BLKGETSIZE64 _IOR(0x12,114,size_t)

#ifdef USE_LOG
void log_message(int priority, const char *format, ...)
{
	va_list args;
	static int udev_log = -1;

	if (udev_log == -1) {
		const char *value;

		value = getenv("UDEV_LOG");
		if (value)
			udev_log = log_priority(value);
		else
			udev_log = LOG_ERR;
	}

	if (priority > udev_log)
		return;

	va_start(args, format);
	vsyslog(priority, format, args);
	va_end(args);
}
#endif

static void vid_log(int priority, const char *file, int line, const char *format, ...)
{
#ifdef USE_LOG
	char log_str[1024];
	va_list args;

	va_start(args, format);
	vsnprintf(log_str, sizeof(log_str), format, args);
	log_str[sizeof(log_str)-1] = '\0';
	log_message(priority, "%s:%i %s", file, line, log_str);
	va_end(args);
#endif
	return;
}

static void set_str(char *to, const char *from, size_t count)
{
	size_t i, j, len;

	/* strip trailing whitespace */
	len = strnlen(from, count);
	while (len && isspace(from[len-1]))
		len--;

	/* strip leading whitespace */
	i = 0;
	while (isspace(from[i]) && (i < len))
		i++;

	j = 0;
	while (i < len) {
		/* substitute multiple whitespace */
		if (isspace(from[i])) {
			while (isspace(from[i]))
				i++;
			to[j++] = '_';
		}
		/* skip chars */
		if (from[i] == '/') {
			i++;
			continue;
		}
		to[j++] = from[i++];
	}
	to[j] = '\0';
}

static int all_probers(volume_id_probe_fn_t probe_fn,
		       struct volume_id *id, uint64_t off, uint64_t size,
		       void *data)
{
	const char *type;

	if (probe_fn(id, off, size) == 0)
		if (volume_id_get_type(id, &type))
			printf("%s\n", type);

	return 0;
}

int main(int argc, char *argv[])
{
	static const struct option options[] = {
		{ "label", 0, NULL, 'l' },
		{ "label-raw", 0, NULL, 'L' },
		{ "uuid", 0, NULL, 'u' },
		{ "type", 0, NULL, 't' },
		{ "export", 0, NULL, 'x' },
		{ "skip-raid", 0, NULL, 's' },
		{ "probe-all", 0, NULL, 'a' },
		{ "help", 0, NULL, 'h' },
		{}
	};

	enum print_type {
		PRINT_EXPORT,
		PRINT_TYPE,
		PRINT_LABEL,
		PRINT_UUID,
		PRINT_LABEL_RAW,
	} print = PRINT_EXPORT;

	struct volume_id *vid = NULL;
	char label_safe[256];
	char label_enc[256];
	char uuid_enc[256];
	uint64_t size;
	int skip_raid = 0;
	int probe_all = 0;
	const char *node;
	int fd;
	const char *label, *uuid, *type, *type_version, *usage;
	int retval;
	int rc = 0;

	logging_init("vol_id");

	/* hook in our debug into libvolume_id */
	volume_id_log_fn = vid_log;

	while (1) {
		int option;

		option = getopt_long(argc, argv, "lLutxsah", options, NULL);
		if (option == -1)
			break;

		switch (option) {
		case 'l':
			print = PRINT_LABEL;
			break;
		case 'L':
			print = PRINT_LABEL_RAW;
			break;
		case 'u':
			print = PRINT_UUID;
			break;
		case 't':
			print = PRINT_TYPE;
			break;
		case 'x':
			print = PRINT_EXPORT;
			break;
		case 's':
			skip_raid = 1;
			break;
		case 'a':
			probe_all = 1;
			break;
		case 'h':
			printf("Usage: vol_id [options] <device>\n"
			    " --export        export key/value pairs\n"
			    " --type          filesystem type\n"
			    " --label         filesystem label\n"
			    " --label-raw     raw label\n"
			    " --uuid          filesystem uuid\n"
			    " --skip-raid     don't probe for raid\n"
			    " --probe-all     find possibly conflicting signatures\n"
			    " --help\n\n");
			goto exit;
		default:
			retval = 1;
			goto exit;
		}
	}

	node = argv[optind];
	if (!node) {
		err("no device");
		fprintf(stderr, "no device\n");
		rc = 1;
		goto exit;
	}

	fd = open(node, O_RDONLY);
	if (fd < 0) {
		fprintf(stderr, "%s: error opening volume\n", node);
		rc = 2;
		goto exit;
	}

	vid = volume_id_open_fd(fd);
	if (vid == NULL) {
		rc = 2;
		goto exit;
	}

	if (ioctl(fd, BLKGETSIZE64, &size) != 0)
		size = 0;
	dbg("BLKGETSIZE64=%llu", (unsigned long long)size);

	/* try to drop all privileges before reading disk content */
	if (getuid() == 0) {
		struct passwd *pw;

		pw = getpwnam("nobody");
		if (pw != NULL && pw->pw_uid > 0 && pw->pw_gid > 0) {
			if (setgroups(0, NULL) != 0 ||
			    setgid(pw->pw_gid) != 0 ||
			    setuid(pw->pw_uid) != 0)
				info("unable to drop privileges: %s\n", strerror(errno));
		}
	}

	if (probe_all) {
		volume_id_all_probers(all_probers, vid, 0, size, NULL);
		goto exit;
	}

	if (skip_raid)
		retval = volume_id_probe_filesystem(vid, 0, size);
	else
		retval = volume_id_probe_all(vid, 0, size);
	if (retval != 0) {
		fprintf(stderr, "%s: unknown volume type\n", node);
		rc = 4;
		goto exit;
	}

	if (!volume_id_get_label(vid, &label) ||
	    !volume_id_get_usage(vid, &usage) ||
	    !volume_id_get_type(vid, &type) ||
	    !volume_id_get_type_version(vid, &type_version) ||
	    !volume_id_get_uuid(vid, &uuid)) {
		rc = 4;
		goto exit;
	}

	set_str(label_safe, label, sizeof(label_safe));
	replace_chars(label_safe, ALLOWED_CHARS_INPUT);

	volume_id_encode_string(label, label_enc, sizeof(label_enc));
	volume_id_encode_string(uuid, uuid_enc, sizeof(uuid_enc));

	switch (print) {
	case PRINT_EXPORT:
		printf("ID_FS_USAGE=%s\n", usage);
		printf("ID_FS_TYPE=%s\n", type);
		printf("ID_FS_VERSION=%s\n", type_version);
		printf("ID_FS_UUID=%s\n", uuid);
		printf("ID_FS_UUID_ENC=%s\n", uuid_enc);
		printf("ID_FS_LABEL=%s\n", label);
		printf("ID_FS_LABEL_ENC=%s\n", label_enc);
		printf("ID_FS_LABEL_SAFE=%s\n", label_safe);
		break;
	case PRINT_TYPE:
		printf("%s\n", type);
		break;
	case PRINT_LABEL:
		if (label_safe[0] == '\0' || strcmp(usage, "raid") == 0) {
			rc = 3;
			goto exit;
		}
		printf("%s\n", label_safe);
		break;
	case PRINT_UUID:
		if (uuid_enc[0] == '\0' || strcmp(usage, "raid") == 0) {
			rc = 4;
			goto exit;
		}
		printf("%s\n", uuid_enc);
		break;
	case PRINT_LABEL_RAW:
		if (label[0] == '\0' || strcmp(usage, "raid") == 0) {
			rc = 3;
			goto exit;
		}
		printf("%s\n", label);
		break;
	}

exit:
	if (vid != NULL)
		volume_id_close(vid);

	logging_close();
	return rc;
}
